### PR TITLE
Make 'Observe' method conditional on having all the references resolved

### LINKED ISSUE #62: Managed Reconciler: A resource with unresolved references never gets deleted
### What happened?

Currently if a managed resource has unresolved references, it never gets deleted, because it exits the reconciliation loop early.

### How can we reproduce it?
 - create a resource which has references that doesn't exist (for example a [`Subnetwork`] with a network reference `NetworkRef` that doesn't exist)
 - delete the `Subnetwork` resource
 - observe that `Subnetwork` never gets deleted 

### What environment did it happen in?
Crossplane version: crossplaneio/crossplane-runtime,master@6916b94

[`Subnetwork`]: https://github.com/crossplaneio/stack-gcp/blob/0b412b0f1627321d22e77b4357ada4db6620f01b/apis/compute/v1alpha3/subnetwork_types.go#L70

### PR DESCRIPTION
**Pull request by soorena776**
Signed-off-by: soorena776 <javad@upbound.io>

### Description of your changes

Fixes #62 
### Checklist

I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation], [examples], or [release notes].
- [ ] Updated the RBAC permissions in [`clusterrole.yaml`] to include any new types.

[documentation]: https://github.com/crossplaneio/crossplane/tree/master/docs
[examples]: https://github.com/crossplaneio/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplaneio/crossplane/tree/master/PendingReleaseNotes.md
[`clusterrole.yaml`]: https://github.com/crossplaneio/crossplane/blob/master/cluster/charts/crossplane/templates/clusterrole.yaml

### CODE REVIEW

File: pkg/resource/managed_reconciler_test.go [hunk 1]
```diff
@@ -229,576 +245,6 @@ func TestManagedReconciler(t *testing.T) {
 			},
 			want: want{result: reconcile.Result{RequeueAfter: defaultManagedShortWait}},
 		},
-		"DeletedButResourceExistsAndReclaimRetain": {
```
Comment (soorena776) [inline pkg/resource/managed_reconciler_test.go]: fixing the unittests

**Comment by muvaf** [comment]
I think the issue here is a more general one; we shouldn't set the finalizer unless we are sure the resource actually exists, either creating or in any form. See https://github.com/crossplaneio/crossplane-runtime/issues/63

**Comment by soorena776** [comment]
@muvaf I saw #63 and I agree. This fix however is focused on where ReferenceResolver should be placed in the Reconcile loop, as it currently always exits the loop early if a reference is not resolved

**Comment by muvaf** [comment]
> // If the references are not resolved the resource cannot exist, assuming
// that the reference fields are immutable.

I am not sure about both of the statements above hold true in our use cases.

On another note, I'm a bit hesitant to make a change on managed reconciler this late to the 0.4 release. How about moving reference resolve block to after `if meta.WasDeleted(managed)` block to solve this specific issue if it's a really pressing one?

**Comment by negz** [comment]
I think we should avoid fixing this for 0.4, which we plan to ship today. My rationale is:

* Fixing it would require last minute changes to the managed resource reconciler, which is used by well over half of our controllers these days. The blast radius of unintended effects on 0.4 would be large.
* Fixing the managed resource reconciler won't fix the issue for the Kubernetes cluster, bucket, and Azure Redis controllers (assuming they are also affected), which do not use the managed resource reconciler.
* I believe it's possible to work around the issue (albeit manually) by editing it to either remove the finalizer, or remove the offending references.
* I'm not certain the change proposed by this PR is the best path forward.

I suggest we consider the bug raised by #62 together with that raised by #63 holistically and see if we can come up with a bug fix for 0.4.1.

**Comment by negz** [comment]
We ended up taking a different path for this in https://github.com/crossplaneio/crossplane-runtime/pull/70, so I'm going to close this PR. I appreciate the effort coming up with a quick fix, even though we did not run with it for 0.4!

**Pull request closed without merging**